src/TestSuite/Fixture/TransactionStrategy.php

```diff
@@ -0,0 +1,131 @@
+    }
+
+    /**
+     * Alias non test connections to the test ones
+     * so that models reach the test database connections instead.
+     *
+     * @param bool $enableLogging Whether or not to enable query logging.
+     * @return void
+     */
+    protected function aliasConnections(bool $enableLogging): void
+    {
+        $connections = ConnectionManager::configured();
+        ConnectionManager::alias('test', 'default');
+        $map = [];
+        foreach ($connections as $connection) {
+            if ($connection === 'test' || $connection === 'default') {
+                continue;
+            }
+            if (isset($map[$connection])) {
+                continue;
+            }
+            if (strpos($connection, 'test_') === 0) {
+                $map[$connection] = substr($connection, 5);
+            } else {
+                $map['test_' . $connection] = $connection;
+            }
+        }
+        foreach ($map as $testConnection => $normal) {
+            ConnectionManager::alias($testConnection, $normal);
+            $connection = ConnectionManager::get($normal);
+            if ($connection instanceof Connection) {
+                if (!$connection->enableSavePoints(true)) {
+                    throw new RuntimeException(
+                        "Could not enable save points for the `{$normal}` connection. " .
+                            'Your database needs to support savepoints in order to use the ' .
+                            'TransactionStrategy for fixtures.'
+                    );
+                }
+                if ($enableLogging) {
+                    $connection->enableQueryLogging(true);
+                }
+            }
+        }
+    }
+
+    /**
+     * Before each test start a transaction.
+     *
+     * @return void
+     */
+    public function beforeTest(): void
+    {
+        $connections = ConnectionManager::configured();
+        foreach ($connections as $connection) {
+            if (strpos($connection, 'test') !== 0) {
+                continue;
+            }
+            $db = ConnectionManager::get($connection);
+            if ($db instanceof Connection) {
+                $db->begin();
+                $db->createSavePoint('__fixtures__');
+            }
+        }
+    }
+
+    /**
+     * After each test rollback the transaction.
+     *
+     * As long as the application code has balanced BEGIN/ROLLBACK
+     * operations we should end up at a transaction depth of 0
+     * and we will rollback the root transaction started in beforeTest()
+     *
+     * @return void
+     */
+    public function afterTest(): void
+    {
+        $connections = ConnectionManager::configured();
+        foreach ($connections as $connection) {
+            if (strpos($connection, 'test') === false) {
+                continue;
+            }
+            $db = ConnectionManager::get($connection);
+            if ($db instanceof Connection) {
+                $db->rollback(true);
```

Comment:
I'm not very familiar with save points, but why do we create them if we are rolling back to beginning here?

Reply:
Application code could be using transactions as well. Using savepoints lets you have 'nested' transactions that can be rolled back in case of an error or constraint violation.

We want to roll everything back at the end of the test so that the database is 'clean' once again. With a transaction based 'state management' fixture records and rows modified during the test would all occur within a transaction. By rolling back at the end we can 'reset' the database without having to figure out which tables need to be truncated.

Reply:
I see. I think I'm just confused by we need the `__fixtures__` save point. Perhaps this code is too early to understand that workflow.

Reply:
We might not, I'm still figuring things out as well.